cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,214 @@
+
+Agents will wait for CIDs to be received through the watcher, and then allocated them to local endpoints (pods) by retrieving the CIDs from the watcher's store based on pod and namespace labels.
+
+**Namespace watch**
+
+Agents will continue to watch namespaces in order to instantly update security identity labels in the local Endpoint manager’s cache. More info below in Label changes: Namespace.
+
+**Pod startup**
+
+CNI will be blocked if CID doesn't exist in the CID watcher's store. After retries, CNI will fail and pod creation will fail. This is useful to quickly identify issues with CID controller or Operator.
+
+Pod startup latency will be reduced when a CID for pod’s labels doesn’t exist yet. CID creation is time consuming mainly because of waiting for kube-apiserver to create the CID resource in Etcd. With operator managing CIDs, CID creation will happen as soon as Pod object is created, before it's even scheduled. Therefore, Agents will always have CID in the watcher store at the time of pod startup.
+
+## Pod creation process
+
+CID assignment paths on pod creation:
+
+1. Regular path: Pod created -> CID created (Operator) -> Pod scheduled -> Cilium CNI -> CID assigned
+2. Missing global ID path: Pod created -> Pod scheduled -> Cilium CNI -> Fail to find CID in watcher Store -> Retry/Fail CNI -> Pod creation is retried by kubelet.
+
+The diagram below covers high level actions that happen when a pod is created. From pod scheduled to pod ready, and to CEP and CID data being received by all nodes in the cluster. CEP batching into CES is included, but that’s an optional step based on configuration. Comparison before (red arrows) and after (blue arrows) changes.
+
+![](../cilium/images/CFP-27752-pod-creation-process.png)
+
+
+CEP creation in parallel triggers its endpoint regeneration on the host node and CEP propagation to all other nodes.
+
+Endpoint regeneration happens when:
+
+* An endpoint is created and updated, including pod and namespace label changes (security identity change).
+* CID is created or deleted - all Agents regenerate all of their endpoints.
+* Network policies are changed.
+
+IPCache is updated when remote CEPs/CESs are received.
+
+## Label changes
+
+
+### Pod
+
+Pod/Deployment label changes result in pod recreation, which follows the same pod creation path previously described.
+
+
+### Namespace
+
+When namespace labels change, Operator and Agents get the update at the same time. Agents wait until Operator processes the namespace label change update and creates all the new CIDs. Agents keep retrying to match CIDs for all their local endpoints until they are received through the CID watcher.
+
+
+## Simultaneous CID management compatibility
+
+The only problem with simultaneous CID management (both Operator and Agents are managing CIDs) is when namespace labels are changed. The only bad thing that happens is that there is CID duplication. This is an existing problem already, that doesn't cause disruption.
```

Comment:
nit: I found this paragraph confusing. I think the same information could be conveyed with simpler text?

```suggestion
Prior to this feature, all agents managed CIDs and operators did not manage CIDs. When upgrading Cilium to enable this feature, we expect users to enable both agents and operators to manage CIDs. As long as Agents are managing CIDs, the CID duplication issue may continue to occur. Once the upgrade is completed, users can disable managing CIDs from the agents.
```

Reply:
Applied